avaframe/com1DFAPy/com1DFAtools.py

```diff
@@ -1307,48 +1343,165 @@ def getNormalArray(x, y, Nx, Ny, Nz, csz):
     return nx, ny, nz
 
 
-def getNormalVect(z, csz, num=4):
+def getNormalMesh(z, csz, num=4):
     n, m = np.shape(z)
     Nx = np.ones((n, m))
     Ny = np.ones((n, m))
     Nz = np.ones((n, m))
     # first and last row, first and last column are inacurate
     if num == 4:
+        # filling the inside of the matrix
         # normal calculation with 4 triangles
-        # (Zl - Zr) * csz
+        # (Zl - Zr) / csz
         Nx[1:n-1, 1:m-1] = (z[1:n-1, 0:m-2] - z[1:n-1, 2:m]) / csz
         # (Zd - Zu) * csz
         Ny[1:n-1, 1:m-1] = (z[0:n-2, 1:m-1] - z[2:n, 1:m-1]) / csz
         Nz = 2 * Nz
+        # filling the first col of the matrix
+        # -2*(Zr - Zp) / csz
+        Nx[1:n-1, 0] = - 2*(z[1:n-1, 1] - z[1:n-1, 0]) / csz
+        # (Zd - Zu) / csz
+        Ny[1:n-1, 0] = (z[0:n-2, 0] - z[2:n, 0]) / csz
+        # filling the last col of the matrix
+        # 2*(Zl - Zp) / csz
+        Nx[1:n-1, m-1] = 2*(z[1:n-1, m-2] - z[1:n-1, m-1]) / csz
+        # (Zd - Zu) / csz
+        Ny[1:n-1, m-1] = (z[0:n-2, m-1] - z[2:n, m-1]) / csz
+        # filling the first row of the matrix
+        # (Zl - Zr) / csz
+        Nx[0, 1:m-1] = (z[0, 0:m-2] - z[0, 2:m]) / csz
+        # -2*(Zu - Zp) / csz
+        Ny[0, 1:m-1] = - 2*(z[1, 1:m-1] - z[0, 1:m-1]) / csz
+        # filling the last row of the matrix
+        # (Zl - Zr) / csz
+        Nx[n-1, 1:m-1] = (z[n-1, 0:m-2] - z[n-1, 2:m]) / csz
+        # 2*(Zd - Zp) / csz
+        Ny[n-1, 1:m-1] = 2*(z[n-2, 1:m-1] - z[n-1, 1:m-1]) / csz
+        # filling the corners of the matrix
+        Nx[0, 0] = -(z[0, 1] - z[0, 0]) / csz
+        Ny[0, 0] = -(z[1, 0] - z[0, 0]) / csz
+        Nz[0, 0] = 1
+        Nx[n-1, 0] = -(z[n-1, 1] - z[n-1, 0]) / csz
+        Ny[n-1, 0] = (z[n-2, 0] - z[n-1, 0]) / csz
+        Nz[n-1, 0] = 1
+        Nx[0, m-1] = (z[0, m-2] - z[0, m-1]) / csz
+        Ny[0, m-1] = -(z[1, m-1] - z[0, m-1]) / csz
+        Nz[0, m-1] = 1
+        Nx[n-1, m-1] = (z[n-1, m-2] - z[n-1, m-1]) / csz
+        Ny[n-1, m-1] = (z[n-2, m-1] - z[n-1, m-1]) / csz
+        Nz[n-1, m-1] = 1
```

Comment:
Is this the best way? It looks very "matlaby" (for a lack of a better word) and it is unclear what exactly is happening. 

If this is the best way to do it, at least group Nz, Ny etc....

Reply:
If you find a better solution I take but so far that is the best I found